### ISSUE TITLE
Integration of POSTGIS

### ISSUE THREAD

**Issue by Kiruel** [serverpod/serverpod#2439]
**Is your feature request related to a problem? Please describe.**
I would like to be able to perform some locations with the ORM. For example [`FT_Intersect`](https://postgis.net/docs/ST_Intersects.html) from POSTGIS. To be able to manipulate vector data.

**Describe the solution you'd like**
For example I would like to do this:

```dart
await Map.db.find(
      session,
      where: (p) =>
          p.point.intersect(GeographyPoint(longitude: 2.55, latitude: 48.55)),
    );
```
Where `point` is a geometry (could be `POINT`, `LINESTRING`, `POLYGON` etc...). And `intersect()` took a geography too. More here: https://postgis.net/docs/ST_Intersects.html

We need something in the yaml file model too like:

```yaml
class: Map
table: map
fields:
  point: geographyPoint
```

All geometry types could be:
```
geographyPolygon
geographyPoint
geographyLinestring
geographyGeometryCollection
```
We need to change the docker image too for postgres to add POSTGIS extension:

Pull the docker image via:
`docker pull postgis/postgis:14-3.4-alpine`

Then change the `docker-compose.yaml` file:
```yaml
services:
  postgres:
    image: postgis/postgis:14-3.4-alpine
```

**Describe alternatives you've considered**
The plan is to add couple geometry and functions like intersect the iterate to add all the function from postgis describe here:
https://postgis.net/docs/reference.html

**Additional context**
Inspiration: https://github.com/serverpod/serverpod/pull/1678

I would like to participate in this contribution, I have already taken a look at the serverpod code and structure I'm a bit blocked by the CLI code for now, if you have any suggestions on how to create this MVP, I will be glad to do it. Then iterate for all POSTGIS capability.

**Comment by SandPod** [comment]
Hey @Kiruel!

Thank you for this detailed feature request.
This seems to be a duplicate of https://github.com/serverpod/serverpod/issues/460, but since this is more detailed, I'm more inclined to close the other issue in favour of this one.

I have not worked with POSTGIS previously, but I believe it is an important addition to the ORM.

I have a couple of questions:
1. Does the class `GeographyPoint` (and the other envisioned geometry types) already exist in some other package, or is it intended to be implemented as part of this feature?
2. I take it as a `point` not being the only column-supported geometry envisioned to be supported but probably good as a starting point.
3. In the YAML configuration, do you envision any other customizations or options that would be required for this feature?

## Implementation requirements for point queries
1. Base query set to accommodate common scenarios (method name in implementation might be different):
    - Points ordered by distance to input. (This, together with `limit` should give us k-nearest points)
    - Points strictly within a certain distance.
    - Points that intersect geometry.
2. Index creation —To ensure fast querying, a smart indexing strategy, possibly involving spatial indexes, might be required.
3. Migration support to add and remove columns.
4. Serialization/Deserialization of data types to support database integration and communication with the client.
5. Compatability with unsafe queries to support all unsupported scenarios.
6. Complete documentation of the feature.
7. Tests for all additions associated with the feature.

## Implementation hints
You mentioned being blocked by the CLI code. Can you elaborate on the specific issues or roadblocks you are encountering?

I'll note down a couple top of mind points that need to be modified in order to support this feature.

### New column
You will need to add a new Column type here: https://github.com/serverpod/serverpod/blob/main/packages/serverpod/lib/src/database/concepts/columns.dart

The column will need the new methods this feature requires along with the `Expressions` that are created as a result from calls to these methods.

The `Expressions` then need to be supported in the SQL query builder located [here](https://github.com/serverpod/serverpod/blob/main/packages/serverpod/lib/src/database/sql_query_builder.dart). But hopefully, this works out of the box.

Then you need to ensure that the column is correctly converted to the correct Postgres type by modifying this method https://github.com/serverpod/serverpod/blob/9f9f04fad5c628b148f4d02ad289ed079f3fc886/packages/serverpod/lib/src/database/adapters/postgres/database_connection.dart#L681C20-L681C32.

### YAML support
In order to support the new keywords in the YAML parser you both need to implement the validation and parsing of the keyword.

Both are handled by our analyzer.
Validation is declaratively defined here: https://github.com/serverpod/serverpod/blob/main/tools/serverpod_cli/lib/src/analyzer/models/yaml_definitions/class_yaml_definition.dart

Parsing is implemented here: https://github.com/serverpod/serverpod/blob/main/tools/serverpod_cli/lib/src/analyzer/models/model_parser/model_parser.dart

### Serialization
Serialization and deserialization will need to be implemented as an extension here: https://github.com/serverpod/serverpod/blob/main/packages/serverpod_serialization/lib/src/extensions/serialization_extensions.dart

And also covered in all the methods here: https://github.com/serverpod/serverpod/blob/main/packages/serverpod_serialization/lib/src/serialization.dart#L247

### Generator
In order for the code generator to use the extensions, they need to be added to this list https://github.com/serverpod/serverpod/blob/9f9f04fad5c628b148f4d02ad289ed079f3fc886/packages/serverpod_serialization/lib/src/serialization.dart#L247.

And then handled here: https://github.com/serverpod/serverpod/blob/9f9f04fad5c628b148f4d02ad289ed079f3fc886/tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart#L690

### Migrations
In order for migrations to support the new column type by modifying this yaml file: https://github.com/serverpod/serverpod/blob/main/packages/serverpod/lib/src/models/database/column_type.spy.yaml

Once that is done you need to add an SQL translation to the new column type here: https://github.com/serverpod/serverpod/blob/9f9f04fad5c628b148f4d02ad289ed079f3fc886/tools/serverpod_cli/lib/src/database/extensions.dart#L307

**Comment by Kiruel** [comment]
Hi!
Thanks for the quick response!

We could follow the work here, I already made a draft PR: https://github.com/serverpod/serverpod/pull/2449 and will add my block concern on the PR.

**Comment by lukehutch** [comment]
This isn't really a total dup of #460, but I'll copy the main things that are missing here.

1. #460 is linked to #1678, where I found that there are some significant changes to the Serverpod query DSL that will be needed to support geo queries.
2. #460 also asks for pg_vector support (crucial for fast AI inference queries)
3. #460 also proposes geohashing for the most likely usecase that users will need (finding nearest neighbors geographically).

**Comment by SandPod** [comment]
1. My concern here is mainly around all the other operations that should probably be supported for a package containing geometric objects to support geometric calculations. I'm guessing that users will like to perform common geometric operations in code using the data types we return.

On the other hand, it might be good for us to own the return format. Since this is going to be part of Serverpod Core, we try to avoid introducing any dependencies.

Just as food for thought, here are a couple of packages that support geometric calculations (don't use them):
- https://pub.dev/packages/geobase
- https://pub.dev/packages/advance_math
https://pub.dev/packages/geodart

2. OK! 👍 Sounds like a good plan!

3. OK!

**Comment by cAttte** [comment]
hey @Kiruel! just another user interested in postgis support and i wanted to know if you've been able to progress, or whether you needed any help implementing this feature.

i also wanted to ask if it's possible to use pg in serverpod as it currently stands (e.g. using raw queries), as it's a pretty important feature for us! mainly i'm stuck on defining a geometry column in the serverpod model file

**Comment by SandPod** [comment]
> i also wanted to ask if it's possible to use pg in serverpod as it currently stands (e.g. using raw queries), as it's a pretty important feature for us! mainly i'm stuck on defining a geometry column in the serverpod model file

@cAttte although you can't define models with geometric data in them, it is possible to manually add these to the database and perform queries using raw queries.

**Comment by koukibadr** [comment]
Hi there! I'd be happy to tackle this issue if no one is currently working on it. I'm interested in implementing PostGIS integration for Serverpod and have experience with spatial data operations.

**Comment by SandPod** [comment]
Go ahead @koukibadr 🚀
